components/sync15/src/key_bundle.rs

```
@@ -75,53 +77,35 @@ impl KeyBundle {
     pub fn hmac_string(&self, ciphertext: &[u8]) -> Result<String> {
         Ok(base16::encode_lower(&self.hmac(ciphertext)?))
     }
 
-    pub fn verify_hmac(&self, expected_hmac: &[u8], ciphertext_base64: &str) -> Result<bool> {
-        let computed_hmac = self.hmac(ciphertext_base64.as_bytes())?;
-        // I suspect this is unnecessary for our case, but the rust-openssl docs
-        // want us to use this over == to avoid sidechannels, and who am I to argue?
-        Ok(openssl::memcmp::eq(&expected_hmac, &computed_hmac))
+    pub fn verify_hmac(&self, expected_hmac: &[u8], ciphertext_base64: &str) -> Result<()> {
+        let key = VerificationKey::new(&digest::SHA256, self.hmac_key());
+        Ok(hmac::verify(
+            &key,
+            ciphertext_base64.as_bytes(),
+            expected_hmac,
+        )?)
     }
 
-    pub fn verify_hmac_string(&self, expected_hmac: &str, ciphertext_base64: &str) -> Result<bool> {
-        let computed_hmac = self.hmac(ciphertext_base64.as_bytes())?;
-        // Note: openssl::memcmp::eq panics if the sizes aren't the same. Desktop returns that it
-        // was a verification failure, so we will too.
-        if expected_hmac.len() != 64 {
-            log::warn!("Garbage HMAC verification string: Wrong length");
-            return Ok(false);
-        }
+    pub fn verify_hmac_string(&self, expected_hmac: &str, ciphertext_base64: &str) -> Result<()> {
         // Decode the expected_hmac into bytes to avoid issues if a client happens to encode
         // this as uppercase. This shouldn't happen in practice, but doing it this way is more
         // robust and avoids an allocation.
         let mut decoded_hmac = [0u8; 32];
-
         if base16::decode_slice(expected_hmac, &mut decoded_hmac).is_err() {
             log::warn!("Garbage HMAC verification string: contained non base16 characters");
-            return Ok(false);
```

I am very, very pleased to see this throwing an error rather than returning `OK(false)`, excellent cleanup!